# Problem Statement and Description

In the Union Territory of New Delhi, there are a lot of places for the residents to go out and eat, and there are people who would like to take the opportunity of serving such people by opening his/her own restraunt. Food business always thrives on the quality of the food served, variety of the dishes and area in which it is operting. But, in large cities like New Delhi, it could be difficult to decide where should one open his new restraunt, after which he would have to decide which cuisines to offer and many more things. No worries, these problems could be easily solved by using the data, and applying certain tools on it.

So, the Final Problem is to-

1. Classify different areas in different Clusters
2. Decide where to open a restraunt

# Data Requirements

Since we have established a problem, we have to go ahead and decide what kinds of data would be needed. To solve the problem for deciding places where people usually go to eat, we have to determine that by clustering the neighborhoods into diiferent categories, and then determine which could be a more appropriate place to operate at. Then we would need the 

Important links:

1. https://simple.wikipedia.org/wiki/List_of_districts_in_Delhi
2. https://en.wikipedia.org/wiki/Neighbourhoods_of_Delhi

These are the sources used in identifying areas in Delhi for further segmentation, analysis and finding optimum place for a person to open a restaraunt.

Apart form this data, Foursquare API will be used to gather data regarding the places in each neighborhood, so that a more detailed and thorough analysis could be conducted by clustering different areas according to the types of places present in the locality and proper findings can be shown.

In [122]:
import requests
import json
import pandas as pd
import matplotlib.pyplot as plt

#### Load and explore the data

In [177]:
import numpy as np # library to handle data in a vectorized manner
import pandas as pd
import json 
from geopy.geocoders import Nominatim 
import requests 
from pandas.io.json import json_normalize
import matplotlib.cm as cm
import matplotlib.colors as colors
from bs4 import BeautifulSoup as bs
from sklearn.cluster import KMeans
import sys
import folium 

print('Libraries imported.')

Libraries imported.


In [2]:
url="https://en.wikipedia.org/wiki/Neighbourhoods_of_Delhi"
req=requests.get(url)

In [3]:
soup = bs(req.content,'lxml')

In [4]:
l={}
for item in soup.find_all('li')[:9]:
    l[item.text[2:]]=[]
keys=list(l.keys())

for item in soup.find_all('li')[12:30]:
    l[keys[0]].append(item.text)

for item in soup.find_all('li')[30:57]:
    l[keys[1]].append(item.text)

for item in soup.find_all('li')[57:71]:
    l[keys[2]].append(item.text)

for item in soup.find_all('li')[71:79]:
    l[keys[3]].append(item.text)

for item in soup.find_all('li')[79:89]:
    l[keys[4]].append(item.text)

for item in soup.find_all('li')[89:108]:
    l[keys[5]].append(item.text)

for item in soup.find_all('li')[108:157]:
    l[keys[6]].append(item.text)

for item in soup.find_all('li')[157:178]:
    l[keys[7]].append(item.text)

for item in soup.find_all('li')[178:198]:
    l[keys[8]].append(item.text)

In [6]:
res={'Districts':[],'Neighborhoods':[]}
for key in l.keys():
    dis=key
    for sub in l[key]:
        res['Districts'].append(dis)
        res['Neighborhoods'].append(sub)

In [8]:
pd.DataFrame(res)

Districts       Neighborhoods
0    North West Delhi        Adarsh Nagar
1    North West Delhi         Ashok Vihar
2    North West Delhi             Azadpur
3    North West Delhi              Bawana
4    North West Delhi           Begum Pur
5    North West Delhi               Dhaka
6    North West Delhi        Jahangirpuri
7    North West Delhi              Karala
8    North West Delhi        Keshav Puram
9    North West Delhi       Kingsway Camp
10   North West Delhi       Kohat Enclave
11   North West Delhi          Model Town
12   North West Delhi              Narela
13   North West Delhi          Pitam Pura
14   North West Delhi           Rani Bagh
15   North West Delhi             Rithala
16   North West Delhi     Rohini Sub City
17   North West Delhi       Shalimar Bagh
18        North Delhi        Shakti Nagar
19        North Delhi      Bara Hindu Rao
20        North Delhi       Chandni Chowk
21        North Delhi       Chawri Bazaar
22        North Delhi         Civil Lines
23        North Delhi        Dariba Kalan
24        North Delhi          Ghantewala
25        North Delhi         Gulabi Bagh
26        North Delhi         Kamla Nagar
27        North Delhi       Kashmiri Gate
28        North Delhi             Kotwali
29        North Delhi         Lahori Gate
..                ...                 ...
156  South West Delhi           Najafgarh
157  South West Delhi             Naraina
158  South West Delhi               Palam
159  South West Delhi  Rama Krishna Puram
160  South West Delhi             Rajokri
161  South West Delhi            Rangpuri
162  South West Delhi      Sarojini Nagar
163  South West Delhi           Sagar Pur
164  South West Delhi         Vasant Kunj
165  South West Delhi        Vasant Vihar
166        West Delhi         Ashok Nagar
167        West Delhi          Bali Nagar
168        West Delhi         Dhaula Kuan
169        West Delhi         Fateh Nagar
170        West Delhi           Janakpuri
171        West Delhi         Kirti Nagar
172        West Delhi          Meera Bagh
173        West Delhi          Moti Nagar
174        West Delhi         Nehru Nagar
175        West Delhi        Partap Nagar
176        West Delhi       Paschim Vihar
177        West Delhi         Patel Nagar
178        West Delhi        Punjabi Bagh
179        West Delhi      Rajouri Garden
180        West Delhi       Shivaji Place
181        West Delhi       Tihar Village
182        West Delhi         Tilak Nagar
183        West Delhi         Uttam Nagar
184        West Delhi         Vikas Nagar
185        West Delhi           Vikaspuri

[186 rows x 2 columns]

Total 186 Neighborhoods are there in Delhi

In [30]:
res_add={'Latitude':[],'Longitude':[]}
tot=len(res['Districts'])
r=1
geolocator = Nominatim(user_agent="dl_explorer")
for dis,sub in zip(res["Districts"],res['Neighborhoods']):
    add=sub+" "+dis+" Delhi"
    lat=""
    max_at=0
    while lat=="":
        max_at+=1
        out=str(round(r/tot*100,2))+ " % "
        sys.stdout.write('\r'+out+sub)
        if max_at==10:
            break
        try: 
            location = geolocator.geocode(add)
            lat = location.latitude
            lng = location.longitude
            
        except: continue
    r+=1
    try:
        location = geolocator.geocode(sub+" Delhi")
        lat = location.latitude
        lng = location.longitude
    except:
        lat=0
        lng=0
    
    res_add['Latitude'].append(lat)
    res_add['Longitude'].append(lng)

100.0 % Vikaspuriargenuram

In [51]:
neighborhoods=pd.DataFrame({**res,**res_add})
neighborhoods.drop(neighborhoods[neighborhoods['Latitude']==0].index,axis=0,inplace=True)

neighborhoods

Districts       Neighborhoods   Latitude  Longitude
0    North West Delhi        Adarsh Nagar  28.714401  77.167288
1    North West Delhi         Ashok Vihar  28.699453  77.184826
2    North West Delhi             Azadpur  28.707657  77.175547
3    North West Delhi              Bawana  28.799660  77.032885
5    North West Delhi               Dhaka  28.708698  77.205749
6    North West Delhi        Jahangirpuri  28.725972  77.162658
7    North West Delhi              Karala  28.735140  77.032511
8    North West Delhi        Keshav Puram  28.688926  77.161683
9    North West Delhi       Kingsway Camp  28.698778  77.204626
10   North West Delhi       Kohat Enclave  28.698041  77.140539
11   North West Delhi          Model Town  28.702714  77.193991
12   North West Delhi              Narela  28.842610  77.091835
13   North West Delhi          Pitam Pura  28.703268  77.132250
14   North West Delhi           Rani Bagh  28.685982  77.132524
15   North West Delhi             Rithala  28.720806  77.107181
17   North West Delhi       Shalimar Bagh  28.717453  77.150867
18        North Delhi        Shakti Nagar  28.679790  77.194914
19        North Delhi      Bara Hindu Rao  28.659518  77.205010
20        North Delhi       Chandni Chowk  28.655983  77.232194
21        North Delhi       Chawri Bazaar  28.649265  77.226515
22        North Delhi         Civil Lines  28.676851  77.225030
23        North Delhi        Dariba Kalan  28.654602  77.233379
25        North Delhi         Gulabi Bagh  28.669649  77.194726
26        North Delhi         Kamla Nagar  28.680344  77.202129
27        North Delhi       Kashmiri Gate  28.669977  77.232059
28        North Delhi             Kotwali  28.651718  77.221939
29        North Delhi         Lahori Gate  28.655828  77.238711
30        North Delhi         Nehru Vihar  28.710745  77.221174
31        North Delhi           Paharganj  28.641499  77.214061
32        North Delhi        Pratap Nagar  28.666718  77.198897
..                ...                 ...        ...        ...
154  South West Delhi           Moti Bagh  28.667837  77.181810
155  South West Delhi             Munirka  28.554886  77.171084
156  South West Delhi           Najafgarh  28.612304  76.982391
157  South West Delhi             Naraina  28.622055  77.138644
158  South West Delhi               Palam  28.591893  77.082824
159  South West Delhi  Rama Krishna Puram  28.575561  77.172303
160  South West Delhi             Rajokri  28.513170  77.110766
161  South West Delhi            Rangpuri  28.540007  77.119775
162  South West Delhi      Sarojini Nagar  28.574157  77.195370
164  South West Delhi         Vasant Kunj  28.529249  77.154134
165  South West Delhi        Vasant Vihar  28.560691  77.160791
166        West Delhi         Ashok Nagar  28.636021  77.101822
167        West Delhi          Bali Nagar  28.654138  77.128178
168        West Delhi         Dhaula Kuan  28.591891  77.161703
169        West Delhi         Fateh Nagar  28.631615  77.100878
170        West Delhi           Janakpuri  28.621927  77.087476
171        West Delhi         Kirti Nagar  28.653281  77.141773
172        West Delhi          Meera Bagh  28.658844  77.090848
173        West Delhi          Moti Nagar  28.657859  77.142429
174        West Delhi         Nehru Nagar  28.568511  77.251385
176        West Delhi       Paschim Vihar  28.669578  77.095956
177        West Delhi         Patel Nagar  28.659809  77.156957
178        West Delhi        Punjabi Bagh  28.668945  77.132461
179        West Delhi      Rajouri Garden  28.642152  77.116060
180        West Delhi       Shivaji Place  28.651657  77.121703
181        West Delhi       Tihar Village  28.634636  77.107112
182        West Delhi         Tilak Nagar  28.636548  77.096496
183        West Delhi         Uttam Nagar  28.624847  77.065286
184        West Delhi         Vikas Nagar  28.645015  77.035028
185        West Delhi           Vikaspuri  28.638419  77.070836

[172 rows x 4 columns]

After removing areas for which geo data was unavailable, there are Total 172 Neighborhhods that will be used in further study

In [72]:
address = 'Delhi'

geolocator = Nominatim(user_agent="dl_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of New Delhi are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of New Delhi are 28.6517178, 77.2219388.


In [73]:
Delhi_Map=folium.Map(location=[latitude,longitude],zoom_start=10)
for sub,lat,lng in zip(neighborhoods['Neighborhoods'],neighborhoods['Latitude'],neighborhoods['Longitude']):
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=sub,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(Delhi_Map)
Delhi_Map